# Tabela 6.1 Exemplo de conjunto de dados play

In [27]:
import pandas as pd
import numpy as np
# Tabela 6.1 
dados = {
    "Tempo": [
        "Chuvoso", "Ensolarado", "Ensolarado", "Nublado", "Chuvoso",
        "Chuvoso", "Nublado", "Nublado", "Ensolarado", "Chuvoso",
        "Nublado", "Ensolarado", "Ensolarado", "Chuvoso"
    ],
    "Temperatura": [
        71, 69, 80, 83, 70,
        65, 64, 72, 75, 68,
        81, 85, 72, 75
    ],
    "Umidade": [
        91, 70, 90, 86, 96,
        70, 65, 90, 70, 80,
        75, 85, 95, 80
    ],
    "Vento": [
        "Sim", "Não", "Sim", "Não", "Não",
        "Sim", "Sim", "Sim", "Sim", "Não",
        "Não", "Não", "Não", "Não"
    ],
    "Joga": [
        "Não", "Sim", "Não", "Sim", "Sim",
        "Não", "Sim", "Sim", "Sim", "Sim",
        "Sim", "Não", "Não", "Sim"
    ]
}
df = pd.DataFrame(dados)
print(df)

         Tempo  Temperatura  Umidade Vento Joga
0      Chuvoso           71       91   Sim  Não
1   Ensolarado           69       70   Não  Sim
2   Ensolarado           80       90   Sim  Não
3      Nublado           83       86   Não  Sim
4      Chuvoso           70       96   Não  Sim
5      Chuvoso           65       70   Sim  Não
6      Nublado           64       65   Sim  Sim
7      Nublado           72       90   Sim  Sim
8   Ensolarado           75       70   Sim  Sim
9      Chuvoso           68       80   Não  Sim
10     Nublado           81       75   Não  Sim
11  Ensolarado           85       85   Não  Não
12  Ensolarado           72       95   Não  Não
13     Chuvoso           75       80   Não  Sim


In [25]:
P_1 = (df["Joga"] == "Sim").sum() / df["Joga"].count()
P_2 = (df["Joga"] == "Não").sum() / df["Joga"].count()

In [29]:
H_Joga = - ( P_1 * np.log2(P_1) + P_2 * np.log2(P_2) )
H_Joga # Entropia de Joga

np.float64(0.9402859586706311)

In [34]:
# P(JOGA=SIM | TEMPO=ENSOLARADO)
P_joga_sol = ((df["Joga"] == "Sim") & (df["Tempo"] == "Ensolarado")).sum()
# P(JOGA=NAO | TEMPO=ENSOLARADO)
P_Njoga_sol = ((df["Joga"] == "Não") & (df["Tempo"] == "Ensolarado")).sum()

In [35]:
# ENSORALADO
P_sol = P_joga_sol + P_Njoga_sol

In [37]:
# Entropia de Joga dado Tempo=Ensolarado
H_Joga_sol = - ( P_joga_sol / P_sol * np.log2(P_joga_sol / P_sol) +
                  P_Njoga_sol / P_sol * np.log2(P_Njoga_sol / P_sol) )
H_Joga_sol

np.float64(0.9709505944546686)

In [54]:
# Entropia de Joga dado Tempo=Chuvoso
P_joga_chu = ((df["Joga"] == "Sim") & (df["Tempo"] == "Chuvoso")).sum()
# P(JOGA=NAO | TEMPO=CHUVOSO)
P_Njoga_chu = ((df["Joga"] == "Não") & (df["Tempo"] == "Chuvoso")).sum()
# CHUVOSO
P_chu = P_joga_chu + P_Njoga_chu
# Entropia de Joga dado Tempo=Chuvoso
H_Joga_chu = - ( P_joga_chu / P_chu * np.log2(P_joga_chu / P_chu) +
                  P_Njoga_chu / P_chu * np.log2(P_Njoga_chu / P_chu) )

In [55]:
H_Joga_chu

np.float64(0.9709505944546686)

In [56]:
# Entropia de Joga dado Tempo=Nublado
P_joga_nub = ((df["Joga"] == "Sim") & (df["Tempo"] == "Nublado")).sum()
# P(JOGA=NAO | TEMPO=NUBLADO)
P_Njoga_nub = ((df["Joga"] == "Não") & (df["Tempo"] == "Nublado")).sum()
# NUBLADO
P_nub = P_joga_nub + P_Njoga_nub
# To avoid log(0) issues, let's check for 0 and handle it:
if P_joga_nub == 0 or P_Njoga_nub == 0:
    H_Joga_nub = 0  # If any probability is zero, we assign entropy as 0 (no uncertainty)
else:
    # Entropia de Joga dado Tempo=Nublado
    H_Joga_nub = - (P_joga_nub / P_nub * np.log2(P_joga_nub / P_nub) +
                     P_Njoga_nub / P_nub * np.log2(P_Njoga_nub / P_nub))
# P_Njoga_nub = 0

In [57]:
# Entropia ponderada de Tempo
H_Tempo = (P_sol / df["Tempo"].count()) * H_Joga_sol + \
           (P_chu / df["Tempo"].count()) * H_Joga_chu + \
           (P_nub / df["Tempo"].count()) * H_Joga_nub
H_Tempo

np.float64(0.6935361388961918)

In [52]:
# IG - Information Gain
IG = H_Joga - H_Tempo
IG

np.float64(0.24674981977443933)

In [58]:
print(f"""IG Tempo: {IG:.4f}
H(TEMPO): {H_Tempo:.4f}     
H(Joga|Nublado): {H_Joga_nub:.4f}
H(Joga|Chuvoso): {H_Joga_chu:.4f}
H(Joga|Ensolarado): {H_Joga_sol:.4f}
H(Njoga|Nublado): {P_Njoga_nub:.4f}
H(Njoga|Chuvoso): {P_Njoga_chu:.4f}
H(Njoga|Ensolarado): {P_Njoga_sol:.4f}""")


IG Tempo: 0.2467
H(TEMPO): 0.6935     
H(Joga|Nublado): 0.0000
H(Joga|Chuvoso): 0.9710
H(Joga|Ensolarado): 0.9710
H(Njoga|Nublado): 0.0000
H(Njoga|Chuvoso): 2.0000
H(Njoga|Ensolarado): 3.0000


 IG (Information Gain) de 0.247 significa que ao saber o Tempo (Chuv, Nubl, Sol) você reduz a incerteza sobre se vai ou não jogar em 0.247 bits. Isso representa 26% da entropia de incerteza inicial (H_joga = 0.9492)